### ISSUE TITLE
Stereoscopic rendering : dynamically adapt distance between eyes to distance to object

### ISSUE THREAD

**Issue by brunetton** [buddhi1980/mandelbulber2#513]
Hi everybody. After some months of inactivity on this project; I'm back again :)
(By the way: Sooo happy to see some OpenCl support !)

I started to play with new formulas introduced in 2.14. As usual, I'm exploring fractals using stereo rendering, crossed eyes to see 3D details without 3D glasses. But I really miss an option to **dynamically adapt distance between eyes to distance to object**.

## Steps to reproduce:
- load the classic dodecahedron from toolbar
- in image settings:
  - proportion: 16/9
  - height: 150 (good enough to explore with quick rendering)
  - stereoscopic left-right rendering - crossed eyes ("swap eyes" checked)
    - leave distance between eyes to default: **0.1**
- look to the rendered image squinted eyes (crossed eyes; same technique as stereograms)
  - => this is okay:
![capture du 2018-08-08 14-30-54](https://user-images.githubusercontent.com/86322/43837076-b49165e4-9b17-11e8-8c21-aa9f0b4e72b5.png)
- start to explore: go closer to the fractal
  - the distance between eyes quickly become too big compared to camera <-> fractal distance, and it quickly become impossible to squint eyes to "follow" this hyperstereoscopy. Example after a few steps in direction of the fractal:
![hyperstereoscopy](https://user-images.githubusercontent.com/86322/43837025-7ea92ae8-9b17-11e8-9022-edffeb28a5e6.png)
  - To continue to be able to see, you'll have to continuously manually adapt distance between eyes to fractal distance (in this example, changing 0.1 to 0.025 makes stereoscopic effect great again 
![back to good stereoscopic base](https://user-images.githubusercontent.com/86322/43837252-38eb7230-9b18-11e8-8c56-300ea25658c3.png)

## Proposal

I propose that we could even write in distance between eyes field:
- a fixed distance, like it's the case today
  - example: `0.1`
- a percent number, referring to a "distance" between viewpoint and fractal (see below)
  - example: `3%`
  - in that case, the distance between eyes would be updated each time I move the camera

### Distance to fractal evaluation 

In my (non 3D developer), the main problem is to evaluate a distance between viewpoint and fractal. This distance should be a average distance because in case of a little part coming close to viewpoint we don't want the whole image to be "focused" on the close little part:

![capture du 2018-08-08 15-12-02](https://user-images.githubusercontent.com/86322/43839057-71e64484-9b1d-11e8-95a0-6e18640145de.png)
(in that case I don't want to "focus" on the object on the left)

Distance calculation looks quick to compute (when moving mouse on rendered image, the distance is updates in real time) => maybe an average distance, taking 1000 
measurements on the image and computing an average should be quick enough to compute ?

## Mandelbulber version
mandelbulber2 2.1.4

**Comment by zebastian** [comment]
hi @brunetton,
your suggestion is reasonable and should be no problem from the performance perspective.
I can look into this...

**Comment by buddhi1980** [comment]
It looks very bad in animation. I have tested it many years ago in Mandelbulber v1. Better would be to add button which will calculate eye distance based on actual image. The user can add eye distance to keyframe animation if it's needed.

**Comment by mancoast** [comment]
Can someone explain in simpler terms what this means? Will this make it easier to generate VR content that pops on the 3d tv?

**Comment by buddhi1980** [comment]
Bad looking of variable distance of eyes in VR comes from fact that our eyes are always at the same distance. If you create VR animation you should decide from the beginning what distance between eyes will be used. It should be enough low to work well with planned closed distance to fractal.
